eclair-core/src/main/scala/fr/acinq/eclair/offer/OfferManager.scala

```diff
@@ -0,0 +1,250 @@
+import fr.acinq.eclair.message.{OnionMessages, Postman}
+import fr.acinq.eclair.payment.receive.MultiPartHandler
+import fr.acinq.eclair.payment.receive.MultiPartHandler.{CreateInvoiceActor, ReceivingRoute}
+import fr.acinq.eclair.payment.{Bolt12Invoice, Invoice}
+import fr.acinq.eclair.wire.protocol.CommonCodecs.{bytes32, bytes64, lengthPrefixedFeaturesCodec, millisatoshi, publicKey, timestampSecond, uint64overflow}
+import fr.acinq.eclair.wire.protocol.OfferTypes.{InvoiceRequest, InvoiceTlv, Offer}
+import fr.acinq.eclair.wire.protocol.PaymentOnion.FinalPayload
+import fr.acinq.eclair.wire.protocol.{GenericTlv, MessageOnion, OfferTypes, OnionMessagePayloadTlv, TlvStream}
+import fr.acinq.eclair.{Feature, Features, MilliSatoshi, NodeParams, TimestampMilli, TimestampSecond, randomBytes32}
+import scodec.bits.ByteVector
+import scodec.codecs._
+import scodec.{Attempt, Codec, DecodeResult}
+
+import scala.concurrent.duration.DurationInt
+import scala.util.{Success, Try}
+
+object OfferManager {
+  // @formatter:off
+  sealed trait Command
+  case class RegisterOffer(offer: Offer, nodeKey: PrivateKey, pathId_opt: Option[ByteVector32], handler: ActorRef[HandlerCommand]) extends Command
+  case class DisableOffer(offer: Offer) extends Command
+  case class RequestInvoice(messagePayload: MessageOnion.FinalPayload, postman: ActorRef[Postman.SendMessage]) extends Command
+  case class Payment(replyTo: ActorRef[MultiPartHandler.GetIncomingPaymentActor.Command], paymentHash: ByteVector32, payload: FinalPayload.Blinded) extends Command
+
+  sealed trait HandlerCommand
+  case class HandleInvoiceRequest(replyTo: ActorRef[InvoiceRequestActor.Command], invoiceRequest: InvoiceRequest) extends HandlerCommand
+  case class HandlePayment(replyTo: ActorRef[PaymentActor.Command], offerId: ByteVector32, data: ByteVector) extends HandlerCommand
+  // @formatter:on
+
+  case class RegisteredOffer(offer: Offer, nodeKey: PrivateKey, pathId_opt: Option[ByteVector32], handler: ActorRef[HandlerCommand])
+
+  case class PaymentMetadata(preimage: ByteVector32,
+                             payerKey: PublicKey,
+                             createdAt: TimestampSecond,
+                             quantity: Long,
+                             amount: MilliSatoshi,
+                             features: Features[Feature],
+                             extraData: ByteVector)
+
+  private val metadataCodec: Codec[PaymentMetadata] =
+    (("preimage" | bytes32) ::
+      ("payerKey" | publicKey) ::
+      ("createdAt" | timestampSecond) ::
+      ("quantity" | uint64overflow) ::
+      ("amount" | millisatoshi) ::
+      ("features" | lengthPrefixedFeaturesCodec) ::
+      ("extraData" | bytes)).as[PaymentMetadata]
+
+  case class SignedMetadata(signature: ByteVector64,
+                            offerId: ByteVector32,
+                            metadata: ByteVector) {
+    def verify(publicKey: PublicKey): Try[PaymentMetadata] = Try {
```

Comment:
Please remove all usages of `Try` in this file, all the error cases should be easy to handle cleanly without throwing exceptions.

Reply:
I've removed most of it. I've kept the `Try[Bolt12Invoice]` because that comes from`MultiPartHandler` and removing it would require a major refactoring.